sharktank/sharktank/layers/paged_attention.py

```diff
@@ -590,10 +592,17 @@ def read(
         pipeline_state = self.unwrap_pipelining(pipeline_state)
         page_ids = self.unwrap_pipelining(page_ids)
 
-        return self.caches[pipeline].read(
+        k, v = self.caches[pipeline].read(
             state=pipeline_state, transformer_block_index=block, page_ids=page_ids
         )
 
+        # If running purely pipelined, we need to rewrap the primitive tensors as replicated tensors.
+        if isinstance(state[pipeline], ReplicatedTensor):
+            k = ReplicatedTensor(ts=[k], devices=state[pipeline].devices)
```

Comment:
This should be handled by `self.caches[pipeline]`. If it isn't setting the devices correctly thats an issue when passing to the subcaches..